File: app/src/main/java/com/example/usbcommunicator/MainActivity.java
package com.example.usbcommunicator;

import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.util.Log;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import java.nio.charset.StandardCharsets;

public final class MainActivity extends AppCompatActivity {
    private final IUsbCallback mCallback = new IUsbCallback() {
        public void onConnectionEstablished() {
            TextView tv = findViewById(R.id.textView);
            tv.setText(usbEngine.connectionStatus());
            tv.setTextColor(Color.GREEN);
        }

        public void onDeviceDisconnected() {
            TextView tv = findViewById(R.id.textView);
            tv.setText(usbEngine.connectionStatus());
            tv.setTextColor(Color.RED);
        }

        public void onDataReceived(@Nullable byte[] data, int num) {
            TextView tv = findViewById(R.id.textView2);
            if (data == null) {
                Log.d("App", "Received empty data!");
                return;
            }

            String text = new String(data, 0, num, StandardCharsets.UTF_8);
            Log.d("App", "Received: " + text);
            if (text.length() > 10) {
                text = text.substring(0, 10) + "...";
            }

            tv.setText(text);
        }

    };
    private UsbEngine usbEngine;

    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        usbEngine = new UsbEngine(getApplicationContext(), mCallback);
        this.onNewIntent(this.getIntent());

        this.findViewById(R.id.button0).setOnClickListener(it -> {
            usbEngine.write("0".getBytes());
        });

        this.findViewById(R.id.button1).setOnClickListener(it -> {
            usbEngine.write("1".getBytes());
        });
    }

    protected void onNewIntent(@Nullable Intent intent) {
        super.onNewIntent(intent);
        usbEngine.onNewIntent(intent);
    }

    protected void onResume() {
        super.onResume();
        Intent intent = this.getIntent();
        this.onNewIntent(intent);
    }
}

File: app/src/main/java/com/example/usbcommunicator/MainActivity.kt
package com.example.usbcommunicator

import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity


class MainActivity : AppCompatActivity() {
    private lateinit var usbEngine: UsbEngine;

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        onNewIntent(intent);

        findViewById<Button>(R.id.button0).setOnClickListener {
            usbEngine.write("0".toByteArray())
        }
        findViewById<Button>(R.id.button1).setOnClickListener {
            usbEngine.write("1".toByteArray())

        }
    }

    override fun onNewIntent(intent: Intent?) {
        if (!::usbEngine.isInitialized) {
            usbEngine = UsbEngine(applicationContext, mCallback)
        }
        usbEngine.onNewIntent(intent)
        super.onNewIntent(intent)
    }

    override fun onResume() {
        super.onResume()
        val intent = intent
        onNewIntent(intent)
    }

    private val mCallback: IUsbCallback = object: IUsbCallback  {
        override fun onConnectionEstablished() {
            val tv = findViewById<TextView>(R.id.textView)
            tv.text = usbEngine.connectionStatus()
            tv.setTextColor(Color.GREEN)
        }

        override fun onDeviceDisconnected() {
            val tv = findViewById<TextView>(R.id.textView)
            tv.text = usbEngine.connectionStatus()
            tv.setTextColor(Color.RED);
        }

        override fun onDataReceived(data: ByteArray?, num: Int) {
            val tv = findViewById<TextView>(R.id.textView2)
            if (data != null) {
                Log.d("App", data.decodeToString(endIndex = num))
                var text = data.decodeToString(endIndex = num)
                if (text.length > 10){
                    text = text.substring(0, 10) + "..."
                }
                tv.text = text
            }
        }
    }
}